src/module.c

```diff
@@ -83,10 +142,12 @@ int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
 		return REDISMODULE_ERR;
 	}
 
-	Proc_Register();            // Register procedures.
-	AR_RegisterFuncs();         // Register arithmetic functions.
-	Agg_RegisterFuncs();        // Register aggregation functions.
-	CypherWhitelist_Build();    // Build whitelist of supported Cypher elements.
+	Proc_Register();         // Register procedures.
+	AR_RegisterFuncs();      // Register arithmetic functions.
+	Agg_RegisterFuncs();     // Register aggregation functions.
+	_PrepareModuleGlobals(); // Set up global lock and variables scoped to the entire module.
+	RegisterForkHooks();     // Set up forking logic to prevent bgsave deadlocks.
+	CypherWhitelist_Build(); // Build whitelist of supported Cypher elements.
```

Comment:
Lastly, can this be tested? It would be nice to introduce a flow test which verifies this fix.

Reply:
I've done fairly extensive manual testing, but it would be non-deterministic and a bit expensive (since we need moderately long-running write queries) to automate? In general I like the idea of adding more concurrency tests, and if we had those it would be easy to throw in some bgsave directives as a chaos-monkey kind of approach. For now, though, I can't think of any decent tests to add!